### PR TITLE
Bug 1868475: Adds test for independent Mode Dashboard

### PR DESCRIPTION
**Pull request by a2batic**
Unit test on breakdown card and details card for independent mode.

Signed-off-by: Kanika <kmurarka@redhat.com>

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import { BreakdownCard } from "../../components/independent-dashboard-page/breakdown-card/card";
+import * as React from "react";
+import { DashboardItems, dropDownItems, headerLink } from "../../__mocks__/independent-mode-dashboards";
+import { Dropdown } from "@console/internal/components/utils";
+import DashboardCardHeader from "@console/shared/src/components/dashboard/dashboard-card/DashboardCardHeader";
+import DashboardCardBody from "@console/shared/src/components/dashboard/dashboard-card/DashboardCardBody";
+import { BreakdownCardBody } from "../../components/dashboard-page/storage-dashboard/breakdown-card/breakdown-body";
+import DashboardCardTitle from "@console/shared/src/components/dashboard/dashboard-card/DashboardCardTitle";
+import { HeaderPrometheusViewLink } from "../../components/dashboard-page/storage-dashboard/breakdown-card/breakdown-header";
+
```
Comment (gnehapk) [inline frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:12]: Sort the imports properly

File: frontend/packages/ceph-storage-plugin/src/components/independent-dashboard-page/details-card/card.tsx [hunk 2]
```diff
@@ -30,57 +37,68 @@ const k8sResources: FirehoseResource[] = [
   },
 ];
 
-const DetailsCard = React.memo(
-  withDashboardResources((props) => {
-    const { watchK8sResource, stopWatchK8sResource, resources } = props;
+export const DetailsCard: React.FC<DashboardItemProps> = ({
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/independent-dashboard-page/details-card/card.tsx:40]: Why this change? This does not look like a refactor. More like changes in aesthetics.

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  it('Should render Card body', () => {
+    expect(wrapper.find(DashboardCardBody).exists()).toBe(true);
+  });
+
+  it('Should render Breakdown Card body', () => {
+    expect(wrapper.find(BreakdownCardBody).exists()).toBe(true);
+    expect(wrapper.find(BreakdownCardBody).props().isLoading).toBe(true);
+    expect(wrapper.find(BreakdownCardBody).props().hasLoadError).toBe(false);
+  });
+});
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:68]: These tests are part of Breakdown Card right? Why are we testing them again? We should only test whether Breakdown card is getting rendered or not. If these tests are not there in Breakdown lets add them there rather than here.

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/details-card.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  it('Should render Card Title', () => {
+    expect(wrapper.find(DashboardCardTitle).exists()).toBe(true);
+  });
+
+  it('Should render details properly', () => {
+    expect(wrapper.find('[data-test-id="cluster-name"]').text()).toEqual('foo');
+    expect(wrapper.find('[data-test-id="cluster-subscription"]').text()).toEqual('fooVersion');
+  });
+});
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/details-card.spec.tsx:45]: This again should be part of details card. We can add these tests in console-shared.

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx (hunk 3 above)
Comment (a2batic) [reply frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:68]: @bipuladh these tests are for breakdown card of independent mode. All these component are part of the card directly, looking at the [component](https://github.com/openshift/console/blob/master/frontend/packages/ceph-storage-plugin/src/components/independent-dashboard-page/breakdown-card/card.tsx). 

I believe test related to how breakdown card uses props should be part of the breakdown-card test. But, as part of this card, we should add test related to expected props been sent to breakdown component, they are set and calculated here.

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
```
Comment (vojtechszocs) [inline frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:1]: Why do we have a structure like:

```
$ tree packages/ceph-storage-plugin/src/components/independent-dashboard-page
packages/ceph-storage-plugin/src/components/independent-dashboard-page
|-- breakdown-card
|   `-- card.tsx
|-- details-card
|   `-- card.tsx
|-- status-card
|   `-- card.tsx
`-- utilization-card
    `-- card.tsx
```

I'd suggest to simplify it to:

```
|-- breakdown-card.tsx
|-- details-card.tsx
|-- status-card.tsx
`-- utilization-card.tsx
```

Also, please keep the path convention for better test vs. tested-unit correlation, by placing this spec at

```
src/__tests__/independent-dashboard-page/breakdown-card.spec.tsx
```

Comment (vojtechszocs) [reply frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:1]: I see that you're following a single-test-directory convention in Ceph plugin:

```
$ find packages/ceph-storage-plugin/src -type d -name __tests__
packages/ceph-storage-plugin/src/__tests__
```

in which case it's more important to keep your `__tests__` structure well organized, in my opinion.

File: frontend/packages/ceph-storage-plugin/src/__mocks__/independent-mode-dashboards.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+export const DashboardItems = {
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/ceph-storage-plugin/src/__mocks__/independent-mode-dashboards.ts:1]: Also, I'd name this `dashboardData`

```suggestion
export const dashboardData = {
```

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+} from '../../__mocks__/independent-mode-dashboards';
+import { HeaderPrometheusViewLink } from '../../components/dashboard-page/storage-dashboard/breakdown-card/breakdown-header';
+
+describe('BreakdownCard', () => {
+  let wrapper: ShallowWrapper<DashboardItemProps>;
+
+  beforeEach(() => {
+    wrapper = shallow(
+      <BreakdownCard
+        watchPrometheus={DashboardItems.watchPrometheus as any}
```
Comment (vojtechszocs) [inline frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:25]:
```suggestion
        watchPrometheus={dashboardData.watchPrometheus as any}
```

File: frontend/packages/ceph-storage-plugin/src/__mocks__/independent-mode-dashboards.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  stopWatchK8sResource: 'foo',
+};
+
+export const dropDownItems = {
+  Pods: 'Pods',
+  Projects: 'Projects',
+  'Storage Classes': 'Storage Classes',
+};
+
+export const headerLink =
```
Comment (vojtechszocs) [inline frontend/packages/ceph-storage-plugin/src/__mocks__/independent-mode-dashboards.ts:30]: `headerLink` and `dropDownItems` are treated as expected test data.

Therefore, I'd suggest naming them `expectedDropDownItems` and `expectedHeaderLink` respectively.

File: frontend/packages/ceph-storage-plugin/src/__mocks__/independent-mode-dashboards.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  'topk(20, (sum(kubelet_volume_stats_used_bytes * on (namespace,persistentvolumeclaim) group_left(storageclass, provisioner) (kube_persistentvolumeclaim_info * on (storageclass)  group_left(provisioner) kube_storageclass_info {provisioner=~"(.*rbd.csi.ceph.com)|(.*cephfs.csi.ceph.com)|(ceph.rook.io/block)"})) by (namespace)))';
+
+export const detailsCardData = {
```
Comment (vojtechszocs) [inline frontend/packages/ceph-storage-plugin/src/__mocks__/independent-mode-dashboards.ts:33]: Why not simply move `detailsCardData.resources` into `dashboardData.resources` and remove the `detailsCardData` export?

This way, you'll have a single source of test data (`dashboardData`) that can be used in all relevant tests.

File: frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx (hunk 3 above)
Comment (vojtechszocs) [reply frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:68]: > These tests are part of Breakdown Card right?

It seems that `BreakdownCard` component tested here is OCS / independent mode specific. The code is currently in `src/components/independent-dashboard-page/breakdown-card/card.tsx`.

Similarly, there is e.g. (OCS specific) `UtilizationCard` in `src/components/independent-dashboard-page/utilization-card/card.tsx`.

Both of these are built on top of `DashboardCard` which is a core Console component.

What we're missing here (I think) is a test for the core `DashboardCard` component:

```
git grep 'DashboardCard' *.spec.*
# no results
```

For now, I'm OK with these tests being here. In future, we should refactor & move relevant bits into ^^ spec.

cc @rawagner

Comment (vojtechszocs) [reply frontend/packages/ceph-storage-plugin/src/__tests__/independent-mode/breakdown-card.spec.tsx:68]: Once we have a core `DashboardCard` spec, tests for OCS specific components built on top of it should only test the OCS specifics (in terms of their render output or behavior).

**Comment by a2batic** [comment]
@vojtechszocs addressed review comments, please review.

**Comment by a2batic** [comment]
@vojtechszocs addressed review comments, please review.
